File: ReportDataAvailability.r

library(NeuroBlu)
library(tidyr)
library(ggplot2)


listAvailableCohorts()


###First we obtain the patients of the cohort
patientsCohort <- getCohort("test")


######Variables with unique values
##BirthYear

birthYear <- getBirthYear(patientsCohort)
       
###Employment
employment <- getEmployment(patientsCohort)
#Ethnicity
ethnicity <- getEthnicity(patientsCohort)
##Gender
gender <- getGender(patientsCohort)
##Marital
marital <- getMarital(patientsCohort)
##Race
race <- getRace(patientsCohort)      
##State
state <- getState(patientsCohort)
##Education
education <- getEducationYears(patientsCohort)

###Merged table for unique values variables
dfMergedUnique <- merge(birthYear, employment, by = "person_id", all = TRUE)
dfMergedUnique <- merge(dfMergedUnique, ethnicity, by = "person_id", all = TRUE)
dfMergedUnique <- merge(dfMergedUnique, gender, by = "person_id", all = TRUE)
dfMergedUnique <- merge(dfMergedUnique, marital, by = "person_id", all = TRUE)
dfMergedUnique <- merge(dfMergedUnique, race, by = "person_id", all = TRUE)
dfMergedUnique <- merge(dfMergedUnique, state, by = "person_id", all = TRUE)
dfMergedUnique <- merge(dfMergedUnique, education, by = "person_id", all = TRUE)

###We will change "Unknown" and 0 to NA
dfMergedUnique[dfMergedUnique == "Unknown"] <- NA
dfMergedUnique[dfMergedUnique == 0 ] <- NA
##We calculate the NAs in the variables       
dfUniqueAvailability <- sapply(dfMergedUnique, function(x) sum(is.na(x)))
###We obtain the percentage of availability
dfUniqueAvailability <- as.data.frame((((dfUniqueAvailability - length(patientsCohort)) * -1)*100) / length(patientsCohort))                        
#We change the title of the colname to percentage
colnames(dfUniqueAvailability) <- "percentage" 
#We add a column with the variables names for ggplot                               
dfUniqueAvailability$variableName <- rownames(dfUniqueAvailability)                               

#We plot the availability of the unique values                               
p <-ggplot(data=dfUniqueAvailability, aes(x=variableName, y=percentage)) + geom_bar(stat="identity") + theme(axis.text.x = element_text(angle = 45))
p <- p + labs(title="Data Availability", x ="", y = "% of patients with data")
p <- p + theme(plot.title = element_text(hjust = 0.5))
ggsave('barplotUnique.png', p, dpi=300, width = 10, height = 8)                              
                           
#############Longitudinal and non-unique data
###Stressors
stressors <- getStressors(patientsCohort)
##We count the unique patients that have stressors data
stressorsUnique <- length(unique(stressors$person_id))     
##We make a dataframe to store the values                               
dfLongitudinal <- data.frame("Stressors", stressorsUnique)
colnames(dfLongitudinal) <- c("variable_name", "number_of_patients") 

####For the longitudinal data we will calculate two measurements:
#a)Patients with at least 1 observed clinical variable
#b)observed clinical variables / total visits 

##We first obtain the visits
visits <- getVisit(patientsCohort)
##Next we remove the duplicates by person_id, visit_date
visitsUnique <- unique(visits[c("person_id", "visit_date")])                               
###We add a column to dfLongtudinal containing the total number of observations to be
##considered with the visits
dfLongitudinal$total_number <- 0  
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("Visits", length(unique(visitsUnique$person_id)), nrow(visitsUnique))                               
                               
##CGI-I
cgii <- getCGII(patientsCohort) 
                                                       
##We calculate the number of patients with at least one variable observed
##                              
cgii_unique <- length(unique(cgii$person_id))
                               
##We add the number of patients with at least 1 observation and the total number of cgii
##The total_number will be later processed in percentages using the number of visits                               
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("CGII", cgii_unique, nrow(cgii)) 

rm(cgii)                               
##CGI-S                               
cgis <- getCGIS(patientsCohort)                               
                                                      
##We calculate the number of patients with at least one variable observed
##                              
cgis_unique <- length(unique(cgis$person_id))  
##We add the number of patients with at least 1 observation and the total number of cgii                               
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("CGIS", cgis_unique, nrow(cgis)) 
rm(cgis)
                               
###Derived_MADRS
derived_madrs <- getDerivedMADRS(patientsCohort)                              
derived_madrs_unique <- length(unique(derived_madrs$person_id))                           
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("Derived_MADRS", derived_madrs_unique, nrow(derived_madrs))                               
rm(derived_madrs_unique)
###DIAGNOSIS
diagnosis <- getDiagnosis(patientsCohort)
diagnosis_unique <- length(unique(diagnosis$person_id))
###Diagnosis per date 0                               
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("diagnosis", diagnosis_unique, 0)   
rm(diagnosis)
##Drug
drug <- getDrug(patientsCohort)                               
drug_unique <- length(unique(drug$person_id)) 
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("drug", drug_unique, 0) 
rm(drug)                               
##FamilyHistory                               
familyHistory <- getFamilyHistory(patientsCohort)
familyHistory_unique <- length(unique(familyHistory$person_id))
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("familyHistory", familyHistory_unique, 0) 
rm(familyHistory)                               
##GAF
gaf <- getGAF(patientsCohort)                               
gaf_unique <- length(unique(gaf$person_id))     
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("GAF", gaf_unique, nrow(gaf)) 
rm(gaf)  
###MADRS
madrs <- getMADRS(patientsCohort)  
if (nrow(madrs) == 0) {
  dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("MADRS", 0, 0)
} else {
  
  
  dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("MADRS", length(unique(madrs$person_id)), nrow(madrs_unique))
                                                                        
  
}                          
rm(madrs)  
                               
##RSAV
rsav <- getRSAV(patientsCohort)
rsav_unique <- length(unique(rsav$person_id))
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("RSAV", rsav_unique, nrow(rsav)) 
rm(rsav)  

##SAV
sav <- getSAV(patientsCohort)
sav_unique <- length(unique(sav$person_id))
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("SAV", sav_unique, nrow(sav)) 
rm(sav)                               
##SmoothCGIS
smoothCGIS <- getSmoothCGIS(patientsCohort)
smoothCGIS_unique <- length(unique(smoothCGIS$person_id)) 
dfLongitudinal[nrow(dfLongitudinal) + 1,] <- c("SmoothCGIS", smoothCGIS_unique, nrow(smoothCGIS))                               
rm(smoothCGIS)    
                               

###Next we calculate the percentage of data availability                              
dfLongitudinal$percentage_available <- (as.numeric(dfLongitudinal$number_of_patients) * 100) / length(patientsCohort)         
write.csv(dfLongitudinal, "dfLongitudinal.csv")
##Plot
#We plot the availability of the unique values
p2 <-ggplot(data=dfLongitudinal, aes(x=variable_name, y=percentage_available)) + geom_bar(stat="identity") + theme(axis.text.x = element_text(angle = 45))
p2 <- p2 + labs(title="Data Availability", x ="", y = "% of patients with data")
p2 <- p2 + theme(plot.title = element_text(hjust = 0.5))
ggsave('barplotUnique.png', p2, dpi=300, width = 10, height = 8)

File: ReportDataAvailability.py
# Import packages
print("Import packages")
import neuroblu as nb
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
from functools import reduce

# neuroblu.get_cohort() function will allow you to get the patient IDs 
# stored within a particular Cohort Builder project.
# In this example, we are getting patient IDs stored within
# a Cohort Builder Project named 'Cohort For Onboarding Task'. 
print("Get desired cohort")
cohort_patients = nb.get_cohort('Cohort For Onboarding Task') 

#Get variables with unique values in the cohort and store them in a dataframe
print("Get data for variables of interest")
cohort_birth_year = nb.get_birth_year(cohort_patients) #Birth year
cohort_gender = nb.get_gender(cohort_patients) #Gender
cohort_city = nb.get_city(cohort_patients) #City
cohort_education_years = nb.get_education_years(cohort_patients) #Education years
cohort_employment = nb.get_employment(cohort_patients) #Employment
cohort_ethnicity = nb.get_ethnicity(cohort_patients) #Ethnicity
cohort_marital = nb.get_marital(cohort_patients) #Marital status
cohort_race = nb.get_race(cohort_patients) #Race
cohort_state = nb.get_state(cohort_patients) #State

#Create list of dataframes
all_dataframes = [cohort_birth_year, cohort_gender, cohort_city,
                  cohort_education_years, cohort_employment, cohort_ethnicity, 
                  cohort_marital, cohort_race, cohort_state]

#Merge multiple dataframes into one based on 'person_id'
data = reduce(lambda  left,right: pd.merge(left,right,on=['person_id'],
                                            how='outer'), all_dataframes)

#Remove 'person_id' column
data = data.drop(columns=['person_id'])

#Replace Unknown / UNKNOWN for Nans
data = data.replace('Unknown', np.nan)
data = data.replace('UNKNOWN', np.nan)

#Create dictionary with neat names for variables
names={'birth_year':'Year of birth',
      'gender':'Gender',
      'city':'City',
      'years_in_education':'Education years',
      'employment_status':'Employment status',
      'ethnicity':'Ethnicity',
      'marital_status':'Marital status',
      'race':'Race',
      'state':'State'}

#List of columns names in data
colnames = list(data.columns)

#Create table with available data
dct = {'Total_data':{}, 'Missing_data':{}, 'Available_data':{}, 'Percentage_available_data':{}, 'Proportion_available_data':{}}

for i in colnames:
  
  total = 0 #Set total to zero
  total = len(data['%s'%(i)]) #Estimate total data
  
  missing = 0 #Set missing data to zero
  missing = data['%s'%(i)].isnull().sum()
  
  available = 0 #Set available to zero
  available = total - missing #Estimate total available data
  
  per_available = 0 #Set per_available to zero 
  per_available = available / total *100
  
  proportion_available = 0 #Set proportion_available to zero
  proportion_available = available / total
  
  dct['Total_data'][names[(i)]] = '%i'%(total) #Add total data
  dct['Missing_data'][names[(i)]] = '%i'%(missing)
  dct['Available_data'][names[(i)]] = '%i'%(available)
  dct['Percentage_available_data'][names[(i)]] = '%.2f'%(per_available)
  dct['Proportion_available_data'][names[(i)]] = '%.4f'%(proportion_available)

#Transform dictionary into dataframe
df = pd.DataFrame(dct)

#Convert all columns of DataFrame to numeric
df = df.apply(pd.to_numeric)

#Sort df
df = df.sort_values(by=['Available_data'], ascending=True)

#Save df to csv
df.to_csv('Table_Available_Data_NonLongitudinal.csv')

#Plot Data Available 
#Plot figure 
plt.figure(figsize=(7,7))
plt.barh(df.index, df['Available_data']) 

#Plot and axis titles
plt.title('Available data')
plt.ylabel('Variables')
plt.xlabel('Number of patients')

#Customize frame
plt.gca().spines['right'].set_visible(False)
plt.gca().spines['top'].set_visible(False)

#Save plot
plt.savefig('Available_Data_NonLongitudinal.png')

#Plot Percentage of Data Available 
#Plot figure
plt.figure(figsize=(7,7))
plt.barh(df.index, df['Percentage_available_data']) 

#Plot and axis titles
plt.title('Available data')
plt.ylabel('Variables')
plt.xlabel('Percentage of patients (%)')

#Customize frame
plt.gca().spines['right'].set_visible(False)
plt.gca().spines['top'].set_visible(False)

#Save plot
plt.savefig('Percentage_Available_Data_NonLongitudinal.png')
